llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -16539,14 +16570,20 @@ static SDValue performUADDVAddCombine(SDValue A, SelectionDAG &DAG) {
     if (SDValue R = performUADDVAddCombine(A.getOperand(1), DAG))
       return DAG.getNode(ISD::ADD, SDLoc(A), A.getValueType(), R,
                          A.getOperand(0));
+
+  if (SDValue R = DetectZextConcat(A, DAG))
+    return R;
+
   return SDValue();
 }
 
 static SDValue performUADDVCombine(SDNode *N, SelectionDAG &DAG) {
   SDValue A = N->getOperand(0);
   if (A.getOpcode() == ISD::ADD)
     if (SDValue R = performUADDVAddCombine(A, DAG))
-      return DAG.getNode(N->getOpcode(), SDLoc(N), N->getValueType(0), R);
+      return R.getOpcode() == AArch64ISD::UADDLV
```

Comment:
Would it be simpler if DetectZextConcat was in this function too, after this if nest?

Reply:
Just out of curiosity do we know why the original code is so odd, i.e.

`return DAG.getNode(N->getOpcode(), SDLoc(N), N->getValueType(0), R);`

and why we can't just return `R` instead?

Reply:
> Just out of curiosity do we know why the original code is so odd, i.e.
> 
> `return DAG.getNode(N->getOpcode(), SDLoc(N), N->getValueType(0), R);`
> 
> and why we can't just return `R` instead?

That's because for the DetectAddExtract case, we are returning UADDV(UADDLP(x)), so we need to return UADDV(R) basically. But for DetectZextConcat we are only returning UADDLV(concat) so we can just return R.

Reply:
performUADDVAddCombine is recursive on the add. I think that might be where the regressions are coming from.

Reply:
> Would it be simpler if DetectZextConcat was in this function too, after this if nest?

I tried to follow the structure of the other functions. It looked to me like the functions of type perform<pattern>Combine() only call the functions that actually contain the functionality to do the combine. So it seemed neater to keep DetectZextConcat in performUADDVAddCombine rather than performUADDVCombine.

I could create another function for DetectZextConcat to call in performUADDVCombine if performUADDVAddCombine doesn't look like the right place for it, or it complicates the code.

Reply:
> performUADDVAddCombine is recursive on the add. I think that might be where the regressions are coming from.

I see, I didn't consider that. I'll add a different function for DetectZextConcat and see if that fixes some of the regressions.

Reply:
Moving the function seems to have taken care of the regressions.